kotlin/display-kml/src/main/java/com/esri/arcgisruntime/sample/displaykml/MainActivity.kt

```diff
@@ -0,0 +1,169 @@
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_main)
+        progressBar.visibility = View.VISIBLE
+
+        // create a map with the dark gray canvas basemap
+        val map = ArcGISMap(Basemap.Type.DARK_GRAY_CANVAS_VECTOR, 39.0, -98.0, 4);
+        // set the map to the map view
+        mapView.map = map
+        // prompt user to make a KML source selection when the app has loaded
+        map.addDoneLoadingListener {
+            if (map.loadStatus == LoadStatus.LOADED) {
+                Toast.makeText(applicationContext, R.string.user_prompt, Toast.LENGTH_LONG).show()
+                // hide progress bar once map has loaded
+                progressBar.visibility = View.GONE
+            }
+        }
+    }
+
+    /**
+     * Shows progress indicator if the KML Layer is loading, clears all operational layers and adds
+     * the KML layer to the map as an operational layer.
+     *
+     * @param kmlLayer to add to the map
+     */
+    fun display(kmlLayer: KmlLayer) {
+        // show progress indicator if kml dataset is loading
+        progressBar.visibility = View.VISIBLE
+        kmlLayer.addDoneLoadingListener {
+            if (kmlLayer.loadStatus == LoadStatus.LOADED) {
+                progressBar.visibility = View.GONE
+            }
+        }
+        // clear operational layers before adding the KML layer to the map
+        val operationalLayers = mapView.map.operationalLayers
+        operationalLayers.clear()
+        operationalLayers.add(kmlLayer)
+    }
+
+    /**
+     * Display a KML layer from a URL.
+     */
+    fun changeSourceToURL() {
+        // create a KML data set from a URL
+        val kmlDataset = KmlDataset(getString(R.string.noaa_weather_kml_url))
+        // create a KML layer created from the KML data set and display it on the map
+        val kmlLayer = KmlLayer(kmlDataset)
+        display(kmlLayer)
+        // report layers if failed to load
+        reportErrors(kmlLayer, "Failed to load KML layer from URL: ")
+    }
+
+    /**
+     * Display a KML layer from a portal item.
+     */
+    fun changeSourceToPortalItem() {
+        // create a portal to ArcGIS Online
+        val portal = Portal(getString(R.string.arcgis_online_url))
+        // create a portal item from a KML item id
+        val portalItem = PortalItem(portal, getString(R.string.kml_item_id))
+
+        // a KML layer created from an ArcGIS Online portal item
+        val kmlLayer = KmlLayer(portalItem)
+        display(kmlLayer)
+        // report layers if failed to load
+        reportErrors(kmlLayer, "Failed to load KML layer from portal item: ")
+    }
+
+    /**
+     * Display a kml layer from external storage.
+     */
+    fun changeSourceToFileExternalStorage() {
+        // get the data set stored locally in device external storage
+        val file = getExternalFilesDir(null)?.path + getString(R.string.kml_path)
+        val kmlDataset = KmlDataset(file)
```

Comment:
I usually just inline this rather than create a file variable:

```suggestion
        val kmlDataset = KmlDataset(getExternalFilesDir(null)?.path + getString(R.string.kml_path))
```

Reply:
I prefer to keep it as is so that in the case the user hasn't downloaded the file, the `file` variable will report a useful "file not found" type message with the exact path to where the data should be stored in the `reportErrors` function.:

`/storage/emulated/0/Android/data/com.esri.arcgisruntime.samples.displaykml/files/US_State_Capitals.kml File not found`

If `kmlDataset` variable is used in the `reportErrors` function, the path to where the data should be stored doesn't look so clear to me:

`com.esri.arcgisruntime.ogc.kml.KmlDataset@901b86d file not found`